### PR TITLE
Last step in removing engine.Table from api/client/*

### PR DESCRIPTION
**Pull request by duglin**
Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

File: api/client/images.go [hunk 1]
```diff
@@ -18,26 +19,26 @@ import (
 )
 
 // FIXME: --viz and --tree are deprecated. Remove them in a future version.
-func (cli *DockerCli) WalkTree(noTrunc bool, images *engine.Table, byParent map[string]*engine.Table, prefix string, printNode func(cli *DockerCli, noTrunc bool, image *engine.Env, prefix string)) {
-	length := images.Len()
+func (cli *DockerCli) WalkTree(noTrunc bool, images *[]*types.Image, byParent map[string]*[]*types.Image, prefix string, printNode func(cli *DockerCli, noTrunc bool, image *types.Image, prefix string)) {
```
Comment (LK4D4) [inline api/client/images.go]: Seems like `*[]*` type here is overkill. As I see those slices not changed during `WalkTree`.

File: api/common.go [hunk 2]
```diff
@@ -80,6 +83,61 @@ func DisplayablePorts(ports *engine.Table) string {
...
 }
 
+type ByPrivatePort []types.Port
+
+func (r ByPrivatePort) Len() int           { return len(r) }
+func (r ByPrivatePort) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
+func (r ByPrivatePort) Less(i, j int) bool { return r[i].PrivatePort < r[j].PrivatePort }
+
+// TODO Rename to DisplayablePorts (remove "New") when engine.Table goes away
+func NewDisplayablePorts(ports *[]types.Port) string {
```
Comment (LK4D4) [inline api/common.go]: Same here, no need in pointer to slice.

**Comment by LK4D4** [comment]
@duglin So, is this last step? Because I see all over code that we can do something when we get rid of `engine.Table`.

**Comment by duglin** [comment]
@LK4D4 
re: *[]*  - I didn't use it because things might change (as you noted, they don't), I used it because I didn't want to create copies on each call. I figured pointers to structs would be faster/less overhead. 

re: last step - this just removes it from the CLI. There are still other spots on the daemon where engine.Table is used and I was going to try to get to those next.  Goal: remove engine.Table if we can.

**Comment by LK4D4** [comment]
Slice is already reference type as well as map.

**Comment by duglin** [comment]
@LK4D4 ah, ok.  See if the latest is better.

**Comment by duglin** [comment]
@LK4D4 any more comments on this one?

File: daemon/list.go [hunk 3]
```diff
@@ -126,15 +135,15 @@ func (daemon *Daemon) Containers(job *engine.Job) error {
 			return nil
 		}
 		displayed++
-		out := &engine.Env{}
-		out.SetJson("Id", container.ID)
-		out.SetList("Names", names[container.ID])
+		newC := *new(types.Container)
```
Comment (LK4D4) [inline daemon/list.go]: I don't understand this trick. Is it just `types.Container{ID: container.ID, Names: names[container.ID]}` ?

File: graph/list.go [hunk 4]
```diff
@@ -53,7 +62,8 @@ func (s *TagStore) CmdImages(job *engine.Job) error {
 	if err != nil {
 		return err
 	}
-	lookup := make(map[string]*engine.Env)
+
+	lookup := make(map[string]*types.Image)
```
Comment (LK4D4) [inline graph/list.go]: Why you using pointers here if you anyway converting all of them to value below, when construct `images` slice.

**Comment by LK4D4** [comment]
I'm little concerned about using pointers in one places and values in other for same types. Let's use pointers or values everywhere.

**Comment by duglin** [comment]
@LK4D4 I tried to keep the old logic of when to use pointers and when not to, but I probably got it wrong. See if you think the latest looks better.

**Comment by LK4D4** [comment]
viz/tree code still awful, but I think it was like this forever.
LGTM

**Comment by duglin** [comment]
yea, I didn't want to mess with that code for fear of breaking something that our tests didn't catch.
thanks
ping @jfrazelle

**Comment by estesp** [comment]
LGTM

**Pull request merged**